src/librustc/infer/anon_types/mod.rs

```diff
@@ -691,25 +691,41 @@ impl<'a, 'gcx, 'tcx> Instantiator<'a, 'gcx, 'tcx> {
                     // }
                     // ```
                     if let Some(anon_node_id) = tcx.hir.as_local_node_id(def_id) {
-                        let in_definition_scope = match tcx.hir.expect_item(anon_node_id).node {
-                            // impl trait
-                            hir::ItemKind::Existential(hir::ExistTy {
-                                impl_trait_fn: Some(parent),
-                                ..
-                            }) => parent == self.parent_def_id,
-                            // named existential types
-                            hir::ItemKind::Existential(hir::ExistTy {
-                                impl_trait_fn: None,
-                                ..
-                            }) => may_define_existential_type(
-                                tcx,
-                                self.parent_def_id,
-                                anon_node_id,
-                            ),
-                            _ => {
-                                let anon_parent_node_id = tcx.hir.get_parent(anon_node_id);
-                                self.parent_def_id == tcx.hir.local_def_id(anon_parent_node_id)
+                        let parent_def_id = self.parent_def_id;
+                        let def_scope_default = || {
+                            let anon_parent_node_id = tcx.hir.get_parent(anon_node_id);
+                            parent_def_id == tcx.hir.local_def_id(anon_parent_node_id)
+                        };
+                        let in_definition_scope = match tcx.hir.find(anon_node_id) {
+                            Some(hir::map::NodeItem(item)) => match item.node {
+                                // impl trait
+                                hir::ItemKind::Existential(hir::ExistTy {
+                                    impl_trait_fn: Some(parent),
+                                    ..
+                                }) => parent == self.parent_def_id,
+                                // named existential types
+                                hir::ItemKind::Existential(hir::ExistTy {
+                                    impl_trait_fn: None,
+                                    ..
+                                }) => may_define_existential_type(
+                                    tcx,
+                                    self.parent_def_id,
+                                    anon_node_id,
+                                ),
+                                _ => def_scope_default(),
                             },
+                            Some(hir::map::NodeImplItem(item)) => match item.node {
+                                hir::ImplItemKind::Existential(_) => may_define_existential_type(
```

Comment:
yea the syntax is horrible on purpose. It's the only one everyone could agree on that we do *not* want.

Reply:
I added to points to the tracking issue. Dear god I hope they won't get lost. =)

Reply:
IMO it's fine for `helper` to be a defining use of `Item` because it's "in the scope" of `Item`'s  definition (in this case, the trait impl).

Reply:
Was there another discussion on syntax? Two points:

- sometimes there are no constraints: but `type Item;` is of course ambiguous
- "existential" is not really the correct term (mathematically it means "there exists at least one such ..."); really we just mean "indirectly specified"

So possibly `indirect type Item;` or `unspec type Item;` ?

Reply:
This implemented the placeholder syntax from the RFC. Please check out the discussion on the [tracking issue](https://github.com/rust-lang/rust/issues/34511) and follow up RFCs. The syntax has been heavily discussed there and is currently [being RFCed as `type Item = impl Trait;`](https://github.com/rust-lang/rfcs/pull/2515). So if you disagree with the proposed syntax or otherwise have opinions on this topic, bring it up on that RFC.